### PR TITLE
Remove underscore from expensify-common and reduce usage of lodash and jQuery

### PR DESCRIPTION
**Pull request by Skalakid**
Rebased version of https://github.com/Expensify/expensify-common/pull/704 PR, that were made to sign all commits. [Slack thread](https://swmansion.slack.com/archives/C06BDSWLDPB/p1718208445619269?thread_ts=1716410498.094849&cid=C06BDSWLDPB)

The original goal of this PR was to remove `underscore`, `jQuery`, and `lodash` from files that are connected to `ExpensiMark` so we can workletize these functions and enable the `react-native-live-markdown` to run parser on UI thread. We wanted to block usage of these libraries in the whole repo, however without knowing the use cases and without being able to test it inside the private Expensify reposthat uses some of these files, we decided to:
- remove whole `underscore` from repo, since it's quite easy to do without testing
- replace more complex `underscore` fucntions with `lodash` versions because it doesn;t make. sene to re-write them
- remove all these 3 libraries from: `ExpensiMark`, `Logger`, `Src` so we can workletize functions from these files and run them on UI thread in  `react-native-live-markdown`

### Fixed Issues
https://github.com/Expensify/App/issues/42494

# Tests
1. What unit/integration tests cover your change? What autoQA tests cover your change?
1. What tests did you perform that validates your changed worked?

# QA
1. What does QA need to do to validate your changes?
1. What areas to they need to test for regressions?

### CODE REVIEW

**Comment by Skalakid** [comment]
@puneetlath here is a new PR that have all commit signed

**Comment by puneetlath** [comment]
@hungvu193 can you give it a once-over for good measure?

**Comment by hungvu193** [comment]
Sure

File: lib/ExpensiMark.js [hunk 1]
```diff
@@ -445,17 +445,13 @@ export default class ExpensiMark {
...
                         return `<blockquote>${line}</blockquote>`;
-                    }
-
-                    // Use _.map with the named function
-                    resultString = _.map(resultString, wrapWithBlockquote);
+                    });
 
-                    function processString(m) {
-                        // Recursive function to replace nested <blockquote> with ">"
-                        function replaceBlockquotes(text) {
```
Comment (hungvu193) [inline lib/ExpensiMark.js:458]: We don't need this function anymore right?
In this case, I think we can remove below comment:
```
 // Recursive function to replace nested <blockquote> with ">"
```

File: lib/PubSub.jsx [hunk 2]
```diff
@@ -72,8 +72,8 @@ const PubSubModule = {
      * @returns {String}
      */
     once(eventName, callback, optionalScope) {
-        const scope = _.isObject(optionalScope) ? optionalScope : window;
-        const functionToCallOnce = _.once((...args) => callback.apply(scope, args));
+        const scope = typeof optionalScope === 'object' && optionalScope !== null ? optionalScope : window;
```
Comment (hungvu193) [inline lib/PubSub.jsx:75]:
```suggestion
        const scope = Utils.isObject(optionalScope) && optionalScope !== null ? optionalScope : window;
```

File: lib/components/form/element/combobox.js [hunk 3]
```diff
@@ -186,39 +185,39 @@ class Combobox extends React.Component {
...
                     this.reset(false, nextProps.options, nextProps.alreadySelectedOptions);
                 }
-            } else if (!_.isEqual(nextProps.options, this.props.options) || !_.isEqual(nextProps.alreadySelectedOptions, this.props.alreadySelectedOptions)) {
+            } else if (!isEqual(nextProps.options, this.props.options) || !isEqual(nextProps.alreadySelectedOptions, this.props.alreadySelectedOptions)) {
                 this.reset(false, nextProps.options, nextProps.alreadySelectedOptions);
             }
         }
 
-        if (!_.isUndefined(nextProps.openOnInit) && !_.isEqual(nextProps.openOnInit, this.props.openOnInit)) {
+        if (nextProps.openOnInit && !isEqual(nextProps.openOnInit, this.props.openOnInit)) {
```
Comment (hungvu193) [inline lib/components/form/element/combobox.js:214]: To be safe, can we just use `nextProps.openOnInit !== undefined`?

File: lib/components/form/element/combobox.js [hunk 4]
```diff
@@ -186,39 +185,39 @@ class Combobox extends React.Component {
...
             this.setState({
                 isDropdownOpen: nextProps.openOnInit,
             });
         }
 
-        if (!_.isUndefined(nextProps.isReadOnly) && !_.isEqual(nextProps.isReadOnly, this.props.isReadOnly)) {
+        if (nextProps.isReadOnly && !isEqual(nextProps.isReadOnly, this.props.isReadOnly)) {
```
Comment (hungvu193) [inline lib/components/form/element/combobox.js:220]: To be safe, can we just use `nextProps.isReadOnly !== undefined`?

File: lib/components/form/element/combobox.js [hunk 5]
```diff
@@ -542,18 +545,18 @@ class Combobox extends React.Component {
     setValue(val) {
         // We need to look in `this.options` for the matching option because `this.state.options` is a truncated list
         // and might not have every option
-        const optionMatchingVal = _.findWhere(this.options, {value: val});
-        const currentText = get(optionMatchingVal, 'text', '');
+        const optionMatchingVal = this.options.find((option) => option.value === val);
+        const currentText = optionMatchingVal.text || '';
```
Comment (hungvu193) [inline lib/components/form/element/combobox.js:549]:
```suggestion
        const currentText = optionMatchingVal && optionMatchingVal.text || '';
```

File: lib/Func.jsx [hunk 6]
```diff
@@ -68,7 +67,11 @@ function die() {
  * @returns {Array}
  */
 function mapByName(list, methodName) {
-    return _.map(list, (item) => item[methodName].call(item));
+    let arr = list;
```
[the remark this answers is not in the document]
Comment (Skalakid) [reply lib/Func.jsx:70]: `list` can be an object or an array according to the param description above the function

File: lib/components/form/element/combobox.js (hunk 5 above)
Comment (Skalakid) [reply lib/components/form/element/combobox.js:549]: I think we can just change it to: `optionMatchingVal?.text || ''`

**Review by hungvu193 — APPROVED** [review]
LGTM 🎉

**Comment by Skalakid** [comment]
All yours @puneetlath :D

**Review by puneetlath — APPROVED** [review]

**Review by dangrous — APPROVED** [review]
LGTM!

Re: removing those rules, we should maybe call that out publicly, so that people know we should still try to follow them even if the linter says no. What do you think, @puneetlath?

**Comment by blazejkustra** [comment]
@puneetlath @dangrous Who is brave enough to merge this one? 😄

**Comment by puneetlath** [comment]
@dangrous That sounds good. @Skalakid could you make an announcement about that in #expensify-open-source?

**Comment by Skalakid** [comment]
@puneetlath @dangrous I've posted the [announcement](https://expensify.slack.com/archives/C01GTK53T8Q/p1718614402390769) on #expensify-open-source

**Comment by hungvu193** [comment]
Should we also upgrade expensify-common version in EApp?

**Comment by blazejkustra** [comment]
I think we should update in both LM and E/App @hungvu193. It's already merged in LM, [here's the PR](https://github.com/Expensify/react-native-live-markdown/pull/390#event-13213291616).

**Comment by hungvu193** [comment]
Thanks for your information 😄

**Pull request merged**